# MLB Pitchers Multiple Linear Regression - Statsmodels

In [1]:
# import required modules
import pandas as pd
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix
import pandas as pd
from sklearn.metrics import roc_auc_score
from sklearn.metrics import precision_recall_fscore_support
import numpy as np
import statsmodels.api as sm

In [2]:
# load dataset
data = pd.read_csv('../../Capstone/Data/MLB_Ppipeline1.csv', header=0)
data.shape

(31710, 38)

In [3]:
data.columns

Index(['name_common', 'year_ID', 'team_ID', 'lg_ID', 'xRA_sprp_adj',
       'xRA_def_pitcher', 'RS_def_total', 'RpO_replacement', 'teamRpG', 'age',
       'G', 'GS', 'IPouts', 'IPouts_start', 'IPouts_relief', 'RA', 'xRA',
       'PPF', 'PPF_custom', 'xRA_final', 'BIP', 'BIP_perc', 'runs_above_avg',
       'runs_above_avg_adj', 'runs_above_rep', 'GR_leverage_index_avg', 'WAR',
       'oppRpG', 'pyth_exponent', 'waa_win_perc', 'WAA', 'WAA_adj',
       'oppRpG_rep', 'pyth_exponent_rep', 'waa_win_perc_rep', 'WAR_rep',
       'ERA_plus', 'ER_lg'],
      dtype='object')

In [4]:
features=['xRA_sprp_adj', 'G', 'GS', 'RA', 'BIP', 'BIP_perc', 'pyth_exponent', 'WAA_adj', 'oppRpG_rep', 'pyth_exponent_rep']
X=data[features]
y=data.WAR

In [5]:
# split X and y into training and testing sets
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=2019)

In [6]:
X_train.shape

(23782, 10)

In [7]:
X_test.shape

(7928, 10)

In [8]:
y_train.shape

(23782,)

In [9]:
y_test.shape

(7928,)

In [10]:
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    WAR   R-squared (uncentered):                   0.710
Model:                            OLS   Adj. R-squared (uncentered):              0.710
Method:                 Least Squares   F-statistic:                              7766.
Date:                Mon, 22 Jun 2020   Prob (F-statistic):                        0.00
Time:                        02:32:54   Log-Likelihood:                         -24515.
No. Observations:               31710   AIC:                                  4.905e+04
Df Residuals:                   31700   BIC:                                  4.913e+04
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
xRA_sprp_adj          0.2743      0.006     43.002      0.000       0.262       0.287
G                     0.4054      0.011     36.499      0.000       0.384       0.427
GS                   -0.1947      0.009    -22.571      0.000      -0.212      -0.178
RA                   -1.6116      0.016    -98.254      0.000      -1.644      -1.579
BIP                   0.5478      0.026     21.054      0.000       0.497       0.599
BIP_perc              1.0158      0.018     56.934      0.000       0.981       1.051
pyth_exponent        -0.2540      0.008    -33.415      0.000      -0.269      -0.239
WAA_adj               0.2123      0.004     55.628      0.000       0.205       0.220
oppRpG_rep            1.3470      0.042     32.020      0.000       1.265       1.429
pyth_exponent_rep    -0.9672      0.041    -23.695      0.000      -1.047      -0.887
==============================================================================
Omnibus:                     7510.893   Durbin-Watson:                   1.773
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27645.094
Skew:                          -1.157   Prob(JB):                         0.00
Kurtosis:                       6.946   Cond. No.                         43.4
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Equation
WAR = 0.2743(xRA_sprp_adj) + 0.4054(G) - 0.1947(GS) - 1.6116(RA) + 0.5478(BIP) + 1.0158(BIP_perc) - 0.2540(pyth_exponent) + 0.2123(WAA_adj) + 1.3470(oppRpG_rep) - 0.9672(pyth_exponent_rep)